### ISSUE TITLE
[DNN] ONNX Opset 17 Swin Transformer import fails: "Preparation for broadcasting failed"

### ISSUE THREAD

**Issue by nilaysrivastava** [opencv/opencv#28183]
### System Information

OpenCV version: 4.12.0 (python-headless and standard)
Operating System / Platform: macOS (Apple Silicon / M1)
Python version: 3.9 / 3.11
PyTorch Version: 2.2.2 (Exporting Opset 17)

### Detailed description

When attempting to load a standard `torchvision` Swin Transformer (Tiny) exported with ONNX Opset 17, `cv2.dnn.readNet` crashes with a broadcasting error in `NaryEltwiseLayer`.

The error suggests that the element-wise addition (likely in the Attention block's relative position bias) involves tensor shapes that `prepare_for_broadcast_op` cannot reconcile. This appears to be a regression or a gap in support for specific broadcasting rules introduced in Opset 17 for Transformers.

### Error Trace
```text
[ERROR:0@37.497] global onnx_importer.cpp:1039 handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [Add]:(onnx_node!/features/features.1/features.1.1/attn/Add_2) from domain='ai.onnx'

OpenCV(4.12.0) /modules/dnn/src/onnx/onnx_importer.cpp:1058: error: (-2:Unspecified error) in function 'handleNode'
> Node [Add@ai.onnx]:(onnx_node!/features/features.1/features.1.1/attn/Add_2) parse error: 
> OpenCV(4.12.0) /modules/dnn/src/layers/nary_eltwise_layers.cpp:342: error: (-2:Unspecified error) in function 'virtual void cv::dnn::NaryEltwiseLayerImpl::finalize(InputArrayOfArrays, OutputArrayOfArrays)'
> > NaryEltwiseLayer: Preparation for broadcasting failed:
> >     'helper.prepare_for_broadcast_op()' must be 'true'
```

### Steps to reproduce

```python
import torch
import torchvision
import cv2
import numpy as np

def reproduce_issue():
    # 1. Load Swin-T (Tiny)
    print("Loading Swin-T...")
    model = torchvision.models.swin_t(weights=torchvision.models.Swin_T_Weights.IMAGENET1K_V1)
    model.eval()

    # 2. Export to ONNX Opset 17
    print("Exporting to ONNX...")
    dummy_input = torch.randn(1, 3, 224, 224)
    onnx_path = "swin_t_repro.onnx"
    torch.onnx.export(model, dummy_input, onnx_path, 
                      opset_version=17, 
                      input_names=['input'], 
                      output_names=['output'])

    # 3. Trigger Crash in OpenCV
    print("Loading into OpenCV DNN...")
    net = cv2.dnn.readNet(onnx_path)

if __name__ == "__main__":
    reproduce_issue()
```

### Issue submission checklist

- [x] I report the issue, it's not a question
- [x] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [x] I updated to the latest OpenCV version and the issue is still there
- [x] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by nilaysrivastava** [comment]
### Update: Root Cause Isolated
I have successfully reproduced the crash in C++ and debugged the `NaryEltwiseLayer`.
The crash occurs specifically when **broadcasting empty tensors** (dimension size 0).

**The conflicting shapes:**
- Input 0: `[0, 1]`
- Input 1: `[0, 56]`

**Expected Behavior:**
The operation should succeed with broadcasting (1 -> 56) and return an output shape of `[0, 56]`.

**Actual Behavior:**
`helper.prepare_for_broadcast_op()` returns `false`, causing the assertion failure. OpenCV appears to incorrectly invalidate broadcasting when dimensions are 0, even if they are compatible.

**Comment by ruchithakp6364** [comment]
Thanks for the insights and confirmation.

Since the root cause seems to be isolated to the broadcasting logic in NaryEltwiseLayer (specifically prepare_for_broadcast_op() rejecting valid zero-dimension cases), I’ll try to work on a fix that allows compatible broadcasting for empty tensors and avoids the assertion failure.

I’ll also look into adding a minimal regression test to cover this case, if that aligns with the expected behavior.

Please let me know if there are any preferred guidelines or considerations I should be aware of before submitting a PR.

**Comment by asmorkalov** [comment]
@gursimarsingh @abhishek-gola Could you take a look? Could you also try it with 5.x and report status too.

**Comment by ruchithakp6364** [comment]
Thanks for isolating the root cause — that matches what I’m seeing as well.

I looked into `NaryEltwiseLayer::prepare_for_broadcast_op()` and it seems the failure is caused by this early rejection:

    if (this->shapes[k][i] == 0)
        return false;

This invalidates broadcasting whenever a dimension is 0, even when the shapes are compatible
(e.g. [0, 1] ⊕ [0, 56] → [0, 56]), which is valid per ONNX broadcasting rules.

A possible fix would be to **remove this early return** and allow zero-sized dimensions to
flow through the normal broadcasting/consistency checks. Incompatible shapes
(e.g. [0, 3] vs [0, 5]) would still be rejected later by the existing logic.

I’ve verified this resolves the Swin-T ONNX import crash locally.
I can open a PR with:
- the minimal change in `prepare_for_broadcast_op()`
- a regression test covering zero-dimension broadcasting

Let me know if that approach sounds reasonable.

**Comment by nilaysrivastava** [comment]
@ruchithakp6364 Great finding on `prepare_for_broadcast_op`. I independently verified that as well.

**However, be careful:** Simply removing that check is not enough. In my C++ debugging, bypassing that check allows the graph to construct, but it leads to a **Segmentation Fault** during `NaryEltwiseLayer::forward()`.

**The Root Cause of the Segfault:**
The function `findCommonShape` incorrectly calculates the output shape for empty tensors. It sees input dims `0` and `1`, and since `std::max(0, 1) == 1`, it creates an output tensor of size `[..., 1]` instead of `[..., 0]`.
The engine then tries to iterate over 0 elements using a pointer allocated for >0 elements (or vice versa), causing the crash.

**The Complete Fix:**
I have implemented a fix that addresses both:
1.  Allows zero-dims in `prepare_for_broadcast_op`.
2.  Corrects the broadcasting logic in `findCommonShape` to respect empty tensors (0 wins over 1).

I have verified this fix locally with a full C++ reproduction script and the Segfault is gone. I am wrapping this up into a PR now.

**Comment by nilaysrivastava** [comment]
My PR #28188 is now live. It fixes the Segfault in the `NaryEltwiseLayer` engine.

Once that is merged, the `readNet` call will proceed further down the pipeline. 
I noticed during debugging that the next hurdle will be in `onnx_importer.cpp` (it throws an assertion on zero-dim dynamic shapes). 
I am happy to tackle that importer fix in a separate follow-up PR once the engine fix is verified/merged.
